### ISSUE TITLE
Using RandomIdentitySampler sampler, Cross Entropy doesn't converge.

### ISSUE THREAD

**Issue by SkytreeRom** [KaiyangZhou/deep-person-reid#101]
Hi, Kaiyang,
Firstly, I want to thank for your great work.
When I try to use RandomIdentitySampler sampler to train PCB_P6 model with Cross Entropy loss and Triplet loss, I encounter strange things.  The xent loss doesn't converge and vibrates periodically. Every period in the figure stands for one epoch training procedure.
![newplot](https://user-images.githubusercontent.com/22659133/50957355-008a8100-14f9-11e9-9eef-a2e01cd82460.png)
My training command like this:
`python train_imgreid_xent_htri.py \
        --root /DATASET/data \
        --source-names market1501 \
        --target-names market1501 \
        --train-sampler 'RandomIdentitySampler' \
        --print-freq 50 \
        --num-instances 2 \ 
        --lambda-xent 1.0 \
        --lambda-htri 0.05 \
        --height 256 \
        --width 128 \
        --optim amsgrad \
        --lr 0.0003 \
        --max-epoch 60 \
        --stepsize 20 40 \
        --train-batch-size 16 \                                                                                                                              
        --test-batch-size 100 \
        --arch pcb_p6 \
        --save-dir log/resnet50-market-xtri \
        --gpu-devices 0,1,2,3
`
Did i miss some important operations?
THANKS

**Comment by KaiyangZhou** [comment]
![loss](https://user-images.githubusercontent.com/22972972/52333441-2706e380-29f5-11e9-956d-f631a5c786cd.jpg)
Just a quick try. This is what I got with resnet50 and num-instances=4 for xent loss. Although the loss oscillated at the early stage, it eventually converged.

**Comment by L1aoXingyu** [comment]
I found a bug in RandomIdentitySampler, the __len__ defined in it is not equal to len(iter(final_idxs)).

**Comment by KaiyangZhou** [comment]
@L1aoXingyu Yes, good spot, that was actually done purposefully. As stated [here](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/samplers.py#L33), `self.length` is just an estimated number so as to avoid re-writing the entire code of generating `final_idxs`. Doing this wouldn't change the training as it only affects the case where you do [`len(dataloader)`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/sampler.py#L172).

You are welcome to provide a better solution.

**Comment by JoinWei-PKU** [comment]
> Hi, Kaiyang,
> Firstly, I want to thank for your great work.
> When I try to use RandomIdentitySampler sampler to train PCB_P6 model with Cross Entropy loss and Triplet loss, I encounter strange things. The xent loss doesn't converge and vibrates periodically. Every period in the figure stands for one epoch training procedure.
> ![newplot](https://user-images.githubusercontent.com/22659133/50957355-008a8100-14f9-11e9-9eef-a2e01cd82460.png)
> My training command like this:
> `python train_imgreid_xent_htri.py \ --root /DATASET/data \ --source-names market1501 \ --target-names market1501 \ --train-sampler 'RandomIdentitySampler' \ --print-freq 50 \ --num-instances 2 \ --lambda-xent 1.0 \ --lambda-htri 0.05 \ --height 256 \ --width 128 \ --optim amsgrad \ --lr 0.0003 \ --max-epoch 60 \ --stepsize 20 40 \ --train-batch-size 16 \ --test-batch-size 100 \ --arch pcb_p6 \ --save-dir log/resnet50-market-xtri \ --gpu-devices 0,1,2,3 `
> Did i miss some important operations?
> THANKS

Have you solved the problem? I have meet the similar problem.

**Comment by KaiyangZhou** [comment]
@JoinWei-PKU KU 

I have tried the command (`python train_imgreid_xent_htri.py -s market1501 -t market1501 --train-sampler 'RandomIdentitySampler' --height 384`) you provided and below are the loss curves.

htri loss:
![htri_loss](https://user-images.githubusercontent.com/22972972/53014196-ac4fb680-343f-11e9-9e2d-aea035653559.jpg)

xent loss:
![xent_loss](https://user-images.githubusercontent.com/22972972/53014198-ac4fb680-343f-11e9-9cad-2dfab4a38beb.jpg)

I did find that the xent loss oscillates during training, but the jumps were not too large to break the learning. I would guess this is a normal phenomenon because the htri loss explicitly adjusts the feature vectors in a way that somehow violates the classification metric. As the training goes, such effects are diminished.

Regarding training from scratch, xent+htri did not boost the performance in my previous experiments.

Hope this can help. Please let me know if you find any bugs in the code.

**Comment by weiliuxm** [comment]
Thank your for your codes. 

Which model has the better result for  PCB_P6 model ? With Cross Entropy loss and Triplet loss  or just with Cross Entropy loss?

Thank you in advance.

**Comment by KaiyangZhou** [comment]
using cross entropy loss only is fine

**Comment by ZepengW** [comment]
> @L1aoXingyu Yes, good spot, that was actually done purposefully. As stated [here](https://github.com/KaiyangZhou/deep-person-reid/blob/master/torchreid/samplers.py#L33), `self.length` is just an estimated number so as to avoid re-writing the entire code of generating `final_idxs`. Doing this wouldn't change the training as it only affects the case where you do [`len(dataloader)`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/sampler.py#L172).
> 
> You are welcome to provide a better solution.


LOL, finally, localize the bug for training with pytorh-lightning.

This cause train loop finish early than expected.

This further causes pytorch-lightning to be unable to implement automatic evaluation (each n Epoch) and scheduler_step (update at epoch-level).

Just recorded for search and reference by people who have the same problem as me.
